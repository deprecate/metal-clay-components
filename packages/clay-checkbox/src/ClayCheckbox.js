import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayCheckbox.soy.js';

/**
 * Implementation of the Metal Clay Checkbox.
 */
class ClayCheckbox extends Component {
	/**
	 * @inheritDoc
	 */
	syncIndeterminate() {
		this.refs.input.indeterminate = this.indeterminate;
	}
}

/**
 * State definition.
 * @type {!Object}
 * @static
 */
ClayCheckbox.STATE = {
	/**
	 * Flag to indicate if checkbox is checked.
	 * @instance
	 * @memberof ClayCheckbox
	 * @type {?bool}
	 * @default false
	 */
	checked: Config.bool().value(false),

	/**
	 * Flag to indicate if checkbox is disabled.
	 * @instance
	 * @memberof ClayCheckbox
	 * @type {?bool}
	 * @default false
	 */
	disabled: Config.bool().value(false),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayCheckbox
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Flag to indicate if label should be hide.
	 * @instance
	 * @memberof ClayCheckbox
	 * @type {?bool}
	 * @default false
	 */
	hideLabel: Config.bool().value(false),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayCheckbox
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * Flag to indicate if checbkox's checked state is indeterminate.
	 * @instance
	 * @memberof ClayCheckbox
	 * @type {?bool}
	 * @default false
	 */
	indeterminate: Config.bool().value(false),

	/**
	 * Flag to indicate if input should be wrapped to be inline or not.
	 * @instance
	 * @memberof ClayCheckbox
	 * @type {?bool}
	 * @default false
	 */
	inline: Config.bool().value(false),

	/**
	 * Label of the input. Required for accesibility.
	 * @instance
	 * @memberof ClayCheckbox
	 * @type {?string|undefined}
	 * @default undefined
	 */
	label: Config.string(),

	/**
	 * Custom content of the checkbox label. Use it to use your custom html.
	 * If this is used label and hideLabel params are ignored.
	 * @instance
	 * @memberof ClayCheckbox
	 * @type {?html|undefined}
	 * @default undefined
	 */
	labelContent: Config.any(),

	/**
	 * Name to be applied to the element.
	 * @instance
	 * @memberof ClayCheckbox
	 * @type {?string|undefined}
	 * @default undefined
	 */
	name: Config.string(),

	/**
	 * Value to be applied to the element.
	 * @instance
	 * @memberof ClayCheckbox
	 * @type {?string|undefined}
	 * @default undefined
	 */
	value: Config.string(),
};

defineWebComponent('clay-checkbox', ClayCheckbox);

Soy.register(ClayCheckbox, templates);

export {ClayCheckbox};
export default ClayCheckbox;
