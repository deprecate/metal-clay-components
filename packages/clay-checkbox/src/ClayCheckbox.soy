{namespace ClayCheckbox}

/**
 * This renders an icon element based on the Clay Checkbox definition.
 */
{template .render}
	{@param label: html | string}
	{@param? checked: bool}
	{@param? disabled: bool}
	{@param? elementClasses: string}
	{@param? hideLabel: bool}
	{@param? hideWrapper: bool}
	{@param? id: string}
	{@param? indeterminate: bool}
	{@param? name: string}
	{@param? value: string}

	{if $hideWrapper}
		{call .input}
			{param checked: $indeterminate or $checked /}
			{param disabled: $disabled /}
			{param elementClasses: $elementClasses /}
			{param hideLabel: $hideLabel /}
			{param id: $id /}
			{param label: $label /}
			{param name: $name /}
			{param value: $value /}
		{/call}
	{else}
		{let $wrapperClasses kind="text"}
			form-check
			{if $disabled}
				{sp}disabled
			{/if}

			{if $elementClasses}
				{sp}{$elementClasses}
			{/if}
		{/let}

		<div class="{$wrapperClasses}">
			{call .input}
				{param checked: $indeterminate or $checked /}
				{param disabled: $disabled /}
				{param hideLabel: $hideLabel /}
				{param id: $id /}
				{param label: $label /}
				{param name: $name /}
				{param value: $value /}
			{/call}
		</div>
	{/if}
{/template}

/**
 * This renders the checkbox input
 **/
{template .input}
	{@param label: html | string}
	{@param? checked: bool}
	{@param? disabled: bool}
	{@param? elementClasses: string}
	{@param? hideLabel: bool}
	{@param? id: string}
	{@param? name: string}
	{@param? value: string}

	<label class="form-check-label">
		{let $inputAttributes kind="attributes"}
			{if $checked}
				checked="checked"
			{/if}

			{if $disabled}
				{sp}disabled
			{/if}

			class="form-check-input
				{if $elementClasses}
					{sp}{$elementClasses}
				{/if}
			"

			{if $id}
				id="{$id}"
			{/if}

			{if $name}
				name="{$name}"
			{/if}

			{if $value}
				value="{$value}"
			{/if}

			ref="input"
			type="checkbox"
		{/let}

		<input {$inputAttributes}></input>

		{let $spanClasses kind="text"}
			form-check-description
			{if $hideLabel}
				{sp}sr-only
			{/if}
		{/let}

		<span class="{$spanClasses}">{$label}</span>
	</label>
{/template}