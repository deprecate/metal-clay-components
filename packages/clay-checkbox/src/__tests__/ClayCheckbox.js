import ClayCheckbox from '../ClayCheckbox';

let clayCheckbox;

describe('ClayCheckbox', function() {
	afterEach(() => {
		if (clayCheckbox) {
			clayCheckbox.dispose();
		}
	});

	it('should generate default markup', () => {
		clayCheckbox = new ClayCheckbox({
			label: 'My Checkbox',
		});

		expect(clayCheckbox).toMatchSnapshot();
	});

	it('should generate a checked checkbox', () => {
		clayCheckbox = new ClayCheckbox({
			checked: true,
			label: 'My Checkbox',
		});

		expect(clayCheckbox).toMatchSnapshot();
	});

	it('should generate an indeterminate checkbox', () => {
		clayCheckbox = new ClayCheckbox({
			indeterminate: true,
			label: 'My Checkbox',
		});

		expect(clayCheckbox).toMatchSnapshot();
	});

	it('should generate a disabled checkbox', () => {
		clayCheckbox = new ClayCheckbox({
			disabled: true,
			label: 'My Checkbox',
		});

		expect(clayCheckbox).toMatchSnapshot();
	});

	it('should generate a checked and disabled checkbox', () => {
		clayCheckbox = new ClayCheckbox({
			checked: true,
			disabled: true,
			label: 'My Checkbox',
		});

		expect(clayCheckbox).toMatchSnapshot();
	});

	it('should generate a checkbox with hidden label', () => {
		clayCheckbox = new ClayCheckbox({
			hideLabel: true,
			label: 'My Checkbox',
		});

		expect(clayCheckbox).toMatchSnapshot();
	});

	it('should generate an inline checkbox', () => {
		clayCheckbox = new ClayCheckbox({
			inline: true,
			label: 'My Checkbox',
		});

		expect(clayCheckbox).toMatchSnapshot();
	});

	it('should generate a checkbox with custom classes', () => {
		clayCheckbox = new ClayCheckbox({
			elementClasses: 'my-custom-class',
			label: 'My Checkbox',
		});

		expect(clayCheckbox).toMatchSnapshot();
	});

	it('should generate an inline checkbox with classes', () => {
		clayCheckbox = new ClayCheckbox({
			elementClasses: 'my-custom-class',
			inline: true,
			label: 'My Checkbox',
		});

		expect(clayCheckbox).toMatchSnapshot();
	});

	it('should generate a checkbox with custom id', () => {
		clayCheckbox = new ClayCheckbox({
			id: 'myCheckbox',
			label: 'My Checkbox',
		});

		expect(clayCheckbox).toMatchSnapshot();
	});

	it('should generate an inline checkbox with custom id', () => {
		clayCheckbox = new ClayCheckbox({
			id: 'myCheckbox',
			inline: true,
			label: 'My Checkbox',
		});

		expect(clayCheckbox).toMatchSnapshot();
	});

	it('should generate a checkbox with name', () => {
		clayCheckbox = new ClayCheckbox({
			name: 'myCheckbox',
			label: 'My Checkbox',
		});

		expect(clayCheckbox).toMatchSnapshot();
	});

	it('should generate a checkbox with value', () => {
		clayCheckbox = new ClayCheckbox({
			label: 'My Checkbox',
			value: 'mycheckbox',
		});

		expect(clayCheckbox).toMatchSnapshot();
	});
});
