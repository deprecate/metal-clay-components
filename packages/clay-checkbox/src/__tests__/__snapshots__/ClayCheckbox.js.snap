// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayCheckbox should generate a checkbox with custom classes 1`] = `
<div class="custom-control custom-checkbox my-custom-class">
  <label>
    <input class="custom-control-input" ref="input" type="checkbox">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Checkbox</span>
  </label>
</div>
`;

exports[`ClayCheckbox should generate a checkbox with custom id 1`] = `
<div class="custom-control custom-checkbox" id="myCheckbox">
  <label>
    <input class="custom-control-input" ref="input" type="checkbox">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Checkbox</span>
  </label>
</div>
`;

exports[`ClayCheckbox should generate a checkbox with hidden label 1`] = `
<div class="custom-control custom-checkbox">
  <label>
    <input class="custom-control-input" ref="input" type="checkbox">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description sr-only">My Checkbox</span>
  </label>
</div>
`;

exports[`ClayCheckbox should generate a checkbox with name 1`] = `
<div class="custom-control custom-checkbox">
  <label>
    <input class="custom-control-input" name="myCheckbox" ref="input" type="checkbox">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Checkbox</span>
  </label>
</div>
`;

exports[`ClayCheckbox should generate a checkbox with value 1`] = `
<div class="custom-control custom-checkbox">
  <label>
    <input class="custom-control-input" value="mycheckbox" ref="input" type="checkbox">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Checkbox</span>
  </label>
</div>
`;

exports[`ClayCheckbox should generate a checked and disabled checkbox 1`] = `
<div class="custom-control custom-checkbox">
  <label>
    <input checked="" disabled="disabled" class="custom-control-input" ref="input" type="checkbox">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Checkbox</span>
  </label>
</div>
`;

exports[`ClayCheckbox should generate a checked checkbox 1`] = `
<div class="custom-control custom-checkbox">
  <label>
    <input checked="" class="custom-control-input" ref="input" type="checkbox">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Checkbox</span>
  </label>
</div>
`;

exports[`ClayCheckbox should generate a disabled checkbox 1`] = `
<div class="custom-control custom-checkbox">
  <label>
    <input disabled="disabled" class="custom-control-input" ref="input" type="checkbox">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Checkbox</span>
  </label>
</div>
`;

exports[`ClayCheckbox should generate an indeterminate checkbox 1`] = `
<div class="custom-control custom-checkbox">
  <label>
    <input checked="" class="custom-control-input" ref="input" type="checkbox">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Checkbox</span>
  </label>
</div>
`;

exports[`ClayCheckbox should generate an inline checkbox 1`] = `
<div class="custom-control custom-checkbox custom-control-inline">
  <label>
    <input class="custom-control-input" ref="input" type="checkbox">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Checkbox</span>
  </label>
</div>
`;

exports[`ClayCheckbox should generate an inline checkbox with classes 1`] = `
<div class="custom-control custom-checkbox custom-control-inline my-custom-class">
  <label>
    <input class="custom-control-input" ref="input" type="checkbox">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Checkbox</span>
  </label>
</div>
`;

exports[`ClayCheckbox should generate an inline checkbox with custom id 1`] = `
<div class="custom-control custom-checkbox custom-control-inline" id="myCheckbox">
  <label>
    <input class="custom-control-input" ref="input" type="checkbox">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Checkbox</span>
  </label>
</div>
`;

exports[`ClayCheckbox should generate default markup 1`] = `
<div class="custom-control custom-checkbox">
  <label>
    <input class="custom-control-input" ref="input" type="checkbox">
    <span class="custom-control-indicator"></span>
    <span class="custom-control-description">My Checkbox</span>
  </label>
</div>
`;
